pkg/pusher/pusher.go

```diff
@@ -129,35 +125,46 @@ func (s *Service) chunksWorker(warmupTime time.Duration, tracer *tracing.Tracer)
 	}
 
 	push := func(op *Op) {
+		defer func() {
+			wg.Done()
+			<-sem
+		}()
+
 		s.metrics.TotalToPush.Inc()
 		ctx, logger := ctxLogger()
 		startTime := time.Now()
-		wg.Add(1)
-		go func() {
-			defer func() {
-				wg.Done()
-				<-s.sem
-			}()
-			if err := s.pushChunk(ctx, op.Chunk, logger, op.Direct); err != nil {
-				// warning: ugly flow control
-				// if errc is set it means we are in a direct push,
-				// we therefore communicate the error into the channel
-				// otherwise we assume this is a buffered upload and
-				// therefore we repeat().
-				if op.Err != nil {
-					op.Err <- err
+
+		if !op.Direct {
```

Comment:
So we are not returning error for direct upload chunks?

I was thinking of having this check inside the below valid if loop. If the op.Direct is set, you can return the validChunk error on op.Err back instead of going ahead to push. No?